### ISSUE TITLE
Use Deployment Status API(App Status) for App Service

### ISSUE THREAD

**Issue by pamelafox** [Azure/azure-dev#2669]
Currently, when you deploy to App Service, you see the endpoint URI and that azd has successfully finished.

However, oftentimes, App Service is still spinning up the app, and it won't work for another 5-10 minutes. That manifests in two possible ways:

* The App Service default app screen, which suggests that no app has been deployed yet.
* An error screen. This can happen if azd provisioning set an appCommandLine setting that produces an error when the code isn't there yet, like "python3 -m gunicorn main:app", resulting in "main cannot be found".

I suggest adding messaging after deployment that says. "App Service may still be processing your app deployment. If you are seeing a default app screen or error message, consider waiting 10 minutes and checking again. If you are still not seeing your app at that point, check out the App Service deployment logs and system logs for more details."

I usually link users to my debugging tips at http://blog.pamelafox.org/2023/06/tips-for-debugging-flask-deployments-to.html but maybe there's a Learn.com article (or could be an article) with similar content.

cc @btardif @charris-msft

**Comment by pamelafox** [comment]
Note that this should not be needed in the bright beautiful feature when App Service is better able to report its deployment health.

**Comment by savannahostrowski** [comment]
<3 thanks for filing this. I do think it'd be wise for us to print something out at the end of deployment for app service hosted apps so that folks don't go down a debugging rabbit hole when the solution is just to wait 5 minutes. 

@ellismg @rajeshkamal5050 Can we get this prioritized for September?

**Comment by btardif** [comment]
FYI @tulikac 

@savannahostrowski lets sync on this, we have a new API that should help with this scenario.

Basically, the current API treats deployment as just binary copy of bits from local box to azure.

Once binary copy is done, there are still a bunch of things that happen server side before the app starts up.

The new "deployment status" API provides a lot more insight and stateful information for all the things that happen server side after binary copy is complete and before the app is ready to serve the first request.

**Comment by iappala** [comment]
I'm still having the same issue. Tried a completely new resource group with a different location, and first fails on Tiktoken, then  cannot find the rust compiler, but just trying to browse the site, I had the same issue; literally wasted 5 hours, but no use. I would appreciate any help you can give me here. 

Error on new install from scratch:

Note: This error originates from a subprocess and is likely not a problem with pip.
  ERROR: Failed building wheel for aiohttp
  Building wheel for tiktoken (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for tiktoken (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [38 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.macosx-10.9-universal2-cpython-312
      creating build/lib.macosx-10.9-universal2-cpython-312/tiktoken
      copying tiktoken/registry.py -> build/lib.macosx-10.9-universal2-cpython-312/tiktoken
      copying tiktoken/__init__.py -> build/lib.macosx-10.9-universal2-cpython-312/tiktoken
      copying tiktoken/core.py -> build/lib.macosx-10.9-universal2-cpython-312/tiktoken
      copying tiktoken/model.py -> build/lib.macosx-10.9-universal2-cpython-312/tiktoken
      copying tiktoken/load.py -> build/lib.macosx-10.9-universal2-cpython-312/tiktoken
      copying tiktoken/_educational.py -> build/lib.macosx-10.9-universal2-cpython-312/tiktoken
      creating build/lib.macosx-10.9-universal2-cpython-312/tiktoken_ext
      copying tiktoken_ext/openai_public.py -> build/lib.macosx-10.9-universal2-cpython-312/tiktoken_ext
      running egg_info
      writing tiktoken.egg-info/PKG-INFO
      writing dependency_links to tiktoken.egg-info/dependency_links.txt
      writing requirements to tiktoken.egg-info/requires.txt
      writing top-level names to tiktoken.egg-info/top_level.txt
      reading manifest file 'tiktoken.egg-info/SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      warning: no files found matching 'Makefile'
      adding license file 'LICENSE'
      writing manifest file 'tiktoken.egg-info/SOURCES.txt'
      copying tiktoken/py.typed -> build/lib.macosx-10.9-universal2-cpython-312/tiktoken
      running build_ext
      running build_rust
      error: can't find Rust compiler
      
      If you are using an outdated pip version, it is possible a prebuilt wheel is available for this package but pip is not able to install from it. Installing from the wheel would avoid the need for a Rust compiler.
      
      To update pip, run:
      
          pip install --upgrade pip
      
      and then retry package installation.
      
      If you did intend to build this package from source, try installing a Rust compiler from your system package manager and ensure it is on the PATH during installation. Alternatively, rustup (available at https://rustup.rs) is the recommended way to download and update the Rust compiler toolchain.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for tiktoken
Failed to build aiohttp tiktoken
ERROR: Could not build wheels for aiohttp, tiktoken, which is required to install pyproject.toml-based projects
Running "prepdocs.py"
/Users/indra/Documents/Development/OpenAIHHH/./scripts/prepdocs.py:137: SyntaxWarning: invalid escape sequence '\d'
  epilog="Example: prepdocs.py '..\data\*' --storageaccount myaccount --container mycontainer --searchservice mysearch --index myindex -v",
Traceback (most recent call last):


-----------

I do the azd up added the applicationstartcommand : python3 -m gunicorn main:app'

Error logs from Stream Logs:

023-11-14T07:57:47.461974425Z Starting OpenBSD Secure Shell server: sshd.
2023-11-14T07:57:47.600363026Z Site's appCommandLine: python3 -m gunicorn main:app
2023-11-14T07:57:47.610900903Z Launching oryx with: create-script -appPath /home/site/wwwroot -output /opt/startup/startup.sh -virtualEnvName antenv -defaultApp /opt/defaultsite -userStartupCommand 'python3 -m gunicorn main:app'
2023-11-14T07:57:48.005379357Z Could not find build manifest file at '/home/site/wwwroot/oryx-manifest.toml'
2023-11-14T07:57:48.009901290Z Could not find operation ID in manifest. Generating an operation id...
2023-11-14T07:57:48.009916290Z Build Operation ID: 054554a0-5869-4ab1-95e2-0a651a292ba4
2023-11-14T07:57:48.819554849Z Oryx Version: 0.2.20230707.1, Commit: 0bd28e69919b5e8beba451e8677e3345f0be8361, ReleaseTagName: 20230707.1
2023-11-14T07:57:48.920348279Z Writing output script to '/opt/startup/startup.sh'
2023-11-14T07:57:49.142004683Z WARNING: Could not find virtual environment directory /home/site/wwwroot/antenv.
2023-11-14T07:57:49.142059583Z WARNING: Could not find package directory /home/site/wwwroot/__oryx_packages__.
2023-11-14T07:57:53.309930042Z [2023-11-14 07:57:53 +0000] [59] [INFO] Starting gunicorn 20.1.0
2023-11-14T07:57:53.359499900Z [2023-11-14 07:57:53 +0000] [59] [INFO] Listening at: http://0.0.0.0:8000 (59)
2023-11-14T07:57:53.359528000Z [2023-11-14 07:57:53 +0000] [59] [INFO] Using worker: sync
2023-11-14T07:57:53.359534200Z [2023-11-14 07:57:53 +0000] [67] [INFO] Booting worker with pid: 67
2023-11-14T07:57:53.399941693Z [2023-11-14 07:57:53 +0000] [67] [ERROR] Exception in worker process
2023-11-14T07:57:53.399977893Z Traceback (most recent call last):
2023-11-14T07:57:53.399983693Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
2023-11-14T07:57:53.399987593Z     worker.init_process()
2023-11-14T07:57:53.399990993Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/workers/base.py", line 134, in init_process
2023-11-14T07:57:53.399994693Z     self.load_wsgi()
2023-11-14T07:57:53.399997893Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
2023-11-14T07:57:53.400014093Z     self.wsgi = self.app.wsgi()
2023-11-14T07:57:53.400017793Z                 ^^^^^^^^^^^^^^^
2023-11-14T07:57:53.400021093Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/base.py", line 67, in wsgi
2023-11-14T07:57:53.400024393Z     self.callable = self.load()
2023-11-14T07:57:53.400027593Z                     ^^^^^^^^^^^
2023-11-14T07:57:53.400030793Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
2023-11-14T07:57:53.400034093Z     return self.load_wsgiapp()
2023-11-14T07:57:53.400037194Z            ^^^^^^^^^^^^^^^^^^^
2023-11-14T07:57:53.400040494Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
2023-11-14T07:57:53.400043794Z     return util.import_app(self.app_uri)
2023-11-14T07:57:53.400046994Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-14T07:57:53.400050194Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/util.py", line 359, in import_app
2023-11-14T07:57:53.400053594Z     mod = importlib.import_module(module)
2023-11-14T07:57:53.400056794Z           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-14T07:57:53.400059894Z   File "/opt/python/3/lib/python3.11/importlib/__init__.py", line 126, in import_module
2023-11-14T07:57:53.400063194Z     return _bootstrap._gcd_import(name[level:], package, level)
2023-11-14T07:57:53.400066594Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-14T07:57:53.400069694Z   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
2023-11-14T07:57:53.400073494Z   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
2023-11-14T07:57:53.400076794Z   File "<frozen importlib._bootstrap>", line 1140, in _find_and_load_unlocked
2023-11-14T07:57:53.400080294Z ModuleNotFoundError: No module named 'main'
2023-11-14T07:57:53.410917272Z [2023-11-14 07:57:53 +0000] [67] [INFO] Worker exiting (pid: 67)
2023-11-14T07:57:54.038270711Z [2023-11-14 07:57:54 +0000] [59] [INFO] Shutting down: Master
2023-11-14T07:57:54.040413227Z [2023-11-14 07:57:54 +0000] [59] [INFO] Reason: Worker failed to boot.
2023-11-14T07:57:47.951Z INFO  - Pulling image: mcr.microsoft.com/appsvc/middleware:stage3
2023-11-14T07:57:48.214Z INFO  - stage3 Pulling from appsvc/middleware
2023-11-14T07:57:48.235Z INFO  -  Digest: sha256:38bf08d2efff967c2eedd8375eacddca7d3fed318135893dc3d7556f0f532312
2023-11-14T07:57:48.252Z INFO  -  Status: Image is up to date for mcr.microsoft.com/appsvc/middleware:stage3
2023-11-14T07:57:48.303Z INFO  - Pull Image successful, Time taken: 0 Seconds
2023-11-14T07:57:48.604Z INFO  - Starting container for site
2023-11-14T07:57:48.605Z INFO  - docker run -d --expose=8181 --name app-backend-6kgz2kblvhfrq_0_6fe2d5be_middleware -e WEBSITE_CORS_ALLOWED_ORIGINS=https://portal.azure.com,https://ms.portal.azure.com, -e WEBSITE_CORS_SUPPORT_CREDENTIALS=False -e WEBSITE_SITE_NAME=app-backend-6kgz2kblvhfrq -e WEBSITE_AUTH_ENABLED=False -e WEBSITE_ROLE_INSTANCE_ID=0 -e WEBSITE_HOSTNAME=app-backend-6kgz2kblvhfrq.azurewebsites.net -e WEBSITE_INSTANCE_ID=e262ff50b02bc79dfcf96feacb75832df6ff0f14dccff63ca6ee6cb3eb18fb7d -e HTTP_LOGGING_ENABLED=1 mcr.microsoft.com/appsvc/mi2023-11-14T07:57:51.254Z INFO  - Initiating warmup request to container app-backend-6kgz2kblvhfrq_0_6fe2d5be_msiProxy for site app-backend-6kgz2kblvhfrq
2023-11-14T07:57:51.349Z INFO  - Container app-backend-6kgz2kblvhfrq_0_6fe2d5be_msiProxy for site app-backend-6kgz2kblvhfrq initialized successfully and is ready to serve requests.
2023-11-14T07:57:51.358Z INFO  - Initiating warmup request to container app-backend-6kgz2kblvhfrq_0_6fe2d5be for site app-backend-6kgz2kblvhfrq
2023-11-14T07:57:55.187Z ERROR - Container app-backend-6kgz2kblvhfrq_0_6fe2d5be for site app-backend-6kgz2kblvhfrq has exited, failing site start
2023-11-14T07:57:55.188Z INFO  - Initiating warmup request to container app-backend-6kgz2kblvhfrq_0_6fe2d5be_middleware for site app-backend-6kgz2kblvhfrq
2023-11-14T07:58:00.137Z INFO  - Container app-backend-6kgz2kblvhfrq_0_6fe2d5be_middleware for site app-backend-6kgz2kblvhfrq initialized successfully and is ready to serve requests.
2023-11-14T07:58:00.156Z ERROR - Container app-backend-6kgz2kblvhfrq_0_6fe2d5be didn't respond to HTTP pings on port: 8000, failing site start. See container logs for debugging.
2023-11-14T07:58:00.222Z INFO  - Stopping site app-backend-6kgz2kblvhfrq because it failed during startup.
2023-11-14 07:57:58.508 +00:00 [Information] Microsoft.Hosting.Lifetime: Now listening on: http://0.0.0.0:8181
2023-11-14 07:57:58.509 +00:00 [Debug] Microsoft.AspNetCore.Hosting.Diagnostics: Loaded hosting startup assembly Middleware.Host
2023-11-14 07:57:58.509 +00:00 [Information] Microsoft.Hosting.Lifetime: Application started. Press Ctrl+C to shut down.
2023-11-14 07:57:58.519 +00:00 [Information] Microsoft.Hosting.Lifetime: Hosting environment: Production
2023-11-14 07:57:58.519 +00:00 [Information] Microsoft.Hosting.Lifetime: Content root path: /app
2023-11-14 07:57:58.519 +00:00 [Debug] Microsoft.Extensions.Hosting.Internal.Host: Hosting started
2023-11-14 07:57:59.498 +00:00 [Debug] Microsoft.AspNetCore.Server.Kestrel.Connections: Connection id "0HMV4STLNMRPJ" accepted.
2023-11-14 07:57:59.499 +00:00 [Debug] Microsoft.AspNetCore.Server.Kestrel.Connections: Connection id "0HMV4STLNMRPJ" started.
2023-11-14 07:57:59.626 +00:00 [Information] Microsoft.AspNetCore.Hosting.Diagnostics: Request starting HTTP/1.1 GET http://169.254.130.7:8181/robots933456.txt - -
2023-11-14 07:57:59.732 +00:00 [Trace] Middleware: Request came in on normal port, sending to normal target http://169.254.130.4:8000/robots933456.txt
2023-11-14 07:57:59.732 +00:00 [Trace] Middleware: Forwarding request to http://x.x.x.x:8000/robots933456.txt
2023-11-14 07:57:59.931 +00:00 [Error] Middleware: Failed to forward request to http://x.x.x.x:8000. Encountered a System.Net.Http.HttpRequestException exception after 227.243ms with message: Connection refused (169.254.130.4:8000). Check application logs to verify the application is properly handling HTTP traffic.
2023-11-14 07:58:00.126 +00:00 [Debug] Microsoft.AspNetCore.Server.Kestrel.Connections: Connection id "0HMV4STLNMRPJ" completed keep alive response.
2023-11-14 07:58:00.127 +00:00 [Information] Microsoft.AspNetCore.Hosting.Diagnostics: Request finished HTTP/1.1 GET http://169.254.130.7:8181/robots933456.txt - - - 500 0 - 501.1079ms

**Comment by pamelafox** [comment]
Looks like you're trying to build on Python 3.12, but tiktoken has no support for 3.12 yet, can you change to 3.11?

**Comment by iappala** [comment]
> Looks like you're trying to build on Python 3.12, but tiktoken has no support for 3.12 yet, can you change to 3.11?

Sorry for asking a naive question but where do I change that?

**Comment by pamelafox** [comment]
@iappala I saw you had a comment about how to change to 3.11. It depends on how you did the deployment. For example, you can change it in the Portal in Settings, or if you deployed via Bicep, you can change the version to 3.12 there.

**Comment by pamelafox** [comment]
@btardif What's the status of the status API? (harhar) Is it ready for azd developers to be able to use for up/deploy?

**Comment by btardif** [comment]
@tulikac can you help here?

**Comment by tulikac** [comment]
@pamelafox @savannahostrowski - We now have the new Deployment Status API for App Service. I am working with @rajeshkamal5050 and @weikanglim to implement this for Azure Developer CLI. The API has a build and a runtime status - so it will make the deployment experience better by giving appropriate messages.

**Comment by rajeshkamal5050** [comment]
Marking this as a duplicate of #3260 feature request filed by @tulikac
